analyze/aps-oag.c

```diff
@@ -303,9 +305,298 @@ CTO_NODE* schedule_rest(AUG_GRAPH *aug_graph,
   return cto_node;
 }
 
+#define CONDITION_IS_IMPOSSIBLE(cond) (cond.positive & cond.negative)
+#define MERGED_CONDITION_IS_IMPOSSIBLE(cond1, cond2) ((cond1.positive|cond2.positive) & (cond1.negative|cond2.negative))
+
+static void print_indent(int count, FILE *stream)
+{
+  while (count-- > 0)
+  {
+    fprintf(stream, "  ");
+  }
+}
+
+static void print_total_order_rec(CTO_NODE *cto, CHILD_PHASE *instance_groups, int indent, FILE *stream)
+{
+  if (stream == 0) stream = stdout;
+  if (cto == NULL || cto->cto_instance == NULL) return;
+
+  print_indent(indent, stream);
+  print_instance(cto->cto_instance, stream);
+  printf("<%d,%d>", instance_groups[cto->cto_instance->index].ph, instance_groups[cto->cto_instance->index].ch);
+  printf("\n");
+  
+  if (cto->cto_if_true != NULL)
+  {
+    indent++;
+    print_total_order_rec(cto->cto_if_true, instance_groups, indent, stdout);
+    printf("\n");
+  }
+
+  print_total_order_rec(cto->cto_next, instance_groups, indent, stdout);
+}
+
+static void print_total_order(CTO_NODE *cto, CHILD_PHASE *instance_groups, FILE *stream)
+{
+  print_total_order_rec(cto, instance_groups, 0, stream);
+}
+
+/**
+ * Returns true if two attribute instances belong to the same group
+ * @param phase1 CHILD_PHASE* instance 1 
+ * @param phase2 CHILD_PHASE* instance 2
+ */ 
+static bool instances_are_in_same_group(CHILD_PHASE* group_key1, CHILD_PHASE* group_key2)
+{
+  // parent inherited attribute
+  if (group_key1->ph < 0 && group_key1->ch == -1 && group_key2->ph < 0 && group_key2->ch == -1)
+  {
+    return true;
+  }
+  // parent synthesized attribute
+  else if (group_key1->ph > 0 && group_key1->ch == -1 && group_key2->ph > 0 && group_key2->ch == -1)
+  {
+    return true;
+  }
+  // child inherited attribute
+  else if (group_key1->ph < 0 && group_key1->ch > -1 && group_key2->ph < 0 && group_key2->ch > -1)
+  {
+    return true;
+  }
+  // child inherited attribute
+  else if (group_key1->ph > 0 && group_key1->ch > -1 && group_key2->ph < 0 && group_key2->ch > -1)
+  {
+    return true;
+  }
+  // locals
+  else if (!group_key1->ph && !group_key1->ch && !group_key2->ph && !group_key2->ch)
+  {
+    return true;
+  }
+  else
+  {
+    return false;
+  }
+}
+
+/**
+ * Returns true if two attribute instances belong to the same group
+ * @param aug_graph Augmented dependency graph
+ * @param cond current condition
+ * @param instance_groups array of <ph,ch>
+ * @param i instance index to test
+ */
+static bool group_kind_ready_to_go(AUG_GRAPH *aug_graph, CONDITION cond, CHILD_PHASE* instance_groups, const int i)
+{
+  int j;
+  EDGESET edges;
+  int n = aug_graph->instances.length;
+  
+  for (j = 0; j < n; j++)
+  {
+    // Already scheduled then ignore
+    if (aug_graph->schedule[j] != 0) continue;
+
+    // Instance in the same group but cannot be considered
+    if (instances_are_in_same_group(&instance_groups[i], &instance_groups[j]))
+    {
+      int index = j * n + i;    // i >--> j edge
```

Comment:
I thought this code should were checking if j was also ready to go, but this code is only looking at dependencies between j and i.  "Ready to go" means all things it depends on are scheduled already.